File: src/desktop/OpenStickyMemos.Desktop/Services/AuthService.cs
using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Text.Json;
using System.Threading.Tasks;

namespace OpenStickyMemos.Desktop.Services;

public class AuthResponse
{
    public string AccessToken { get; set; } = string.Empty;
    public string RefreshToken { get; set; } = string.Empty;
    public DateTime ExpiresAt { get; set; }
    public UserInfo User { get; set; } = null!;
}

public class UserInfo
{
    public string Id { get; set; } = string.Empty;
    public string Email { get; set; } = string.Empty;
    public string DisplayName { get; set; } = string.Empty;
    public string? AvatarUrl { get; set; }
    public string AuthProvider { get; set; } = string.Empty;
}

public interface IAuthService
{
    string? AccessToken { get; }
    UserInfo? CurrentUser { get; }
    bool IsLoggedIn { get; }
    event Action? AuthChanged;
    Task<bool> LoginWithEmailAsync(string email, string password);
    Task<bool> RegisterWithEmailAsync(string email, string password, string? displayName);
    Task<bool> LoginWithGoogleAsync(string idToken);
    Task<bool> LoginWithMicrosoftAsync(string idToken);
    Task<bool> RefreshTokenAsync();
    Task<bool> TryAutoLoginAsync();
    void Logout();
}

public class AuthService : IAuthService
{
    private readonly HttpClient _http;
    private readonly ISettingsService _settings;
    private readonly ICredentialService _credentials;

    public string? AccessToken { get; private set; }
    public UserInfo? CurrentUser { get; private set; }
    public bool IsLoggedIn => !string.IsNullOrEmpty(AccessToken);
    public event Action? AuthChanged;

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true
    };

    public AuthService(ISettingsService settings, ICredentialService credentials)
    {
        _settings = settings;
        _credentials = credentials;
        _http = new HttpClient { BaseAddress = new Uri(settings.Current.ApiUrl) };
    }

    private static readonly string LogPath = Desktop.App.LogPath;
    private static void LogHttp(string msg)
    {
        try { System.IO.File.AppendAllText(LogPath, $"[{DateTime.Now:HH:mm:ss}] [HTTP] {msg}\n"); } catch { }
    }

    public async Task<bool> LoginWithGoogleAsync(string idToken)
    {
        EnsureBaseAddress();
        return await ExchangeToken(idToken, "Google");
    }

    public async Task<bool> LoginWithMicrosoftAsync(string idToken)
    {
        EnsureBaseAddress();
        return await ExchangeToken(idToken, "Microsoft");
    }

    /// <summary>Asegura que el HttpClient use la URL actual de settings</summary>
    private void EnsureBaseAddress()
    {
        var currentUrl = _settings.Current.ApiUrl;
        if (_http.BaseAddress?.OriginalString != currentUrl)
            _http.BaseAddress = new Uri(currentUrl);
    }

    public async Task<bool> LoginWithEmailAsync(string email, string password)
    {
        EnsureBaseAddress();
        return await EmailAuthAsync("/api/Auth/login", email, password);
    }

    public async Task<bool> RegisterWithEmailAsync(string email, string password, string? displayName)
    {
        EnsureBaseAddress();
        return await EmailAuthAsync("/api/Auth/register", email, password, displayName);
    }

    public async Task<bool> RefreshTokenAsync()
    {
        EnsureBaseAddress();
        var refreshToken = _credentials.GetRefreshToken();
        if (string.IsNullOrEmpty(refreshToken)) return false;

        try
        {
            LogHttp($">> POST /api/Auth/refresh");
            var response = await _http.PostAsJsonAsync("/api/Auth/refresh",
                new { refreshToken });
            var respBody = await response.Content.ReadAsStringAsync();
            LogHttp($"<< {(int)response.StatusCode} {respBody}");

            if (!response.IsSuccessStatusCode) return false;

            var auth = JsonSerializer.Deserialize<AuthResponse>(respBody, JsonOptions);
            if (auth is null) return false;

            ApplyAuth(auth);
            return true;
        }
        catch
        {
            return false;
        }
    }

    public void Logout()
    {
        AccessToken = null;
        CurrentUser = null;
        _credentials.Clear();
        AuthChanged?.Invoke();
    }

    public async Task<bool> TryAutoLoginAsync()
    {
        var token = _credentials.GetAccessToken();
        if (string.IsNullOrEmpty(token)) return false;

        AccessToken = token;
        return await RefreshTokenAsync();
    }

    // ── Private ──

    private async Task<bool> EmailAuthAsync(string endpoint, string email, string password, string? displayName = null)
    {
        try
        {
            var body = displayName is not null
                ? new { email, password, displayName }
                : (object)new { email, password };

            var json = JsonSerializer.Serialize(body);
            var fullUrl = $"{_http.BaseAddress?.OriginalString.TrimEnd('/')}{endpoint}";
            LogHttp($">> POST {fullUrl} {json}");

            var response = await _http.PostAsJsonAsync(endpoint, body);
            var respJson = await response.Content.ReadAsStringAsync();
            LogHttp($"<< {(int)response.StatusCode} {respJson}");

            if (!response.IsSuccessStatusCode)
            {
                System.Windows.MessageBox.Show(
                    $"Respuesta del servidor:\n\nHTTP {(int)response.StatusCode}\n\n{respJson}",
                    "DEBUG Auth",
                    System.Windows.MessageBoxButton.OK,
                    System.Windows.MessageBoxImage.Information);
                return false;
            }

            var auth = JsonSerializer.Deserialize<AuthResponse>(respJson, JsonOptions);
            if (auth is null) return false;

            ApplyAuth(auth);
            return true;
        }
        catch (Exception ex)
        {
            LogHttp($"!! ERROR: {ex.GetType().Name}: {ex.Message}");
            System.Windows.MessageBox.Show(
                $"Error de conexión con el servidor:\n\n{ex.GetType().Name}: {ex.Message}",
                "Error de autenticación",
                System.Windows.MessageBoxButton.OK,
                System.Windows.MessageBoxImage.Error);
            return false;
        }
    }

    private async Task<bool> ExchangeToken(string idToken, string provider)
    {
        try
        {
            var endpoint = provider == "Google" ? "google" : "microsoft";
            LogHttp($">> POST /api/Auth/{endpoint}");
            var response = await _http.PostAsJsonAsync($"/api/Auth/{endpoint}",
                new { idToken, provider });
            var respBody = await response.Content.ReadAsStringAsync();
            LogHttp($"<< {(int)response.StatusCode} {respBody}");

            if (!response.IsSuccessStatusCode) return false;

            var auth = JsonSerializer.Deserialize<AuthResponse>(respBody, JsonOptions);
            if (auth is null) return false;

            ApplyAuth(auth);
            return true;
        }
        catch
        {
            return false;
        }
    }

    private void ApplyAuth(AuthResponse auth)
    {
        AccessToken = auth.AccessToken;
        CurrentUser = auth.User;
        _credentials.Save(auth.AccessToken, auth.RefreshToken,
            JsonSerializer.Serialize(auth.User));

        LogHttp($"ApplyAuth: AccessToken={(AccessToken is not null ? $"{AccessToken[..Math.Min(AccessToken.Length, 30)]}..." : "NULL")}, " +
            $"RefreshToken={(auth.RefreshToken is not null ? $"{auth.RefreshToken[..Math.Min(auth.RefreshToken.Length, 10)]}..." : "NULL")}, " +
            $"User={auth.User?.Email ?? "NULL"}");

        AuthChanged?.Invoke();
    }
}

File: src/web/open-sticky-memos/src/app/core/auth.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap, BehaviorSubject } from 'rxjs';
import { AppConfigService } from './app-config.service';

export interface UserInfo {
  id: string;
  email: string;
  displayName: string;
  avatarUrl: string | null;
  authProvider: string;
}

export interface AuthResponse {
  accessToken: string;
  refreshToken: string;
  expiresAt: string;
  user: UserInfo;
}

@Injectable()
export class AuthService {
  private get apiUrl() { return this.config.apiUrl; }
  private readonly tokenKey = 'osm_access_token';
  private readonly refreshKey = 'osm_refresh_token';
  private readonly userKey = 'osm_user';

  private userSubject = new BehaviorSubject<UserInfo | null>(this.getStoredUser());
  user$ = this.userSubject.asObservable();

  constructor(private http: HttpClient, private config: AppConfigService) {}

  /** Registro con email y contraseña */
  register(email: string, password: string, displayName?: string): Observable<AuthResponse> {
    return this.http
      .post<AuthResponse>(`${this.apiUrl}/auth/register`, { email, password, displayName })
      .pipe(tap((res) => this.storeSession(res)));
  }

  /** Login con email y contraseña */
  login(email: string, password: string): Observable<AuthResponse> {
    return this.http
      .post<AuthResponse>(`${this.apiUrl}/auth/login`, { email, password })
      .pipe(tap((res) => this.storeSession(res)));
  }

  /** Login con Google — envía el id_token al backend */
  loginWithGoogle(idToken: string): Observable<AuthResponse> {
    return this.exchangeToken(idToken, 'Google');
  }

  /** Login con Microsoft — envía el id_token al backend */
  loginWithMicrosoft(idToken: string): Observable<AuthResponse> {
    return this.exchangeToken(idToken, 'Microsoft');
  }

  /** Refresca el access token */
  refreshToken(): Observable<AuthResponse> {
    const refresh = this.getRefreshToken();
    return this.http
      .post<AuthResponse>(`${this.apiUrl}/auth/refresh`, {
        refreshToken: refresh,
      })
      .pipe(tap((res) => this.storeSession(res)));
  }

  /** Logout — limpia la sesión local */
  logout(): void {
    localStorage.removeItem(this.tokenKey);
    localStorage.removeItem(this.refreshKey);
    localStorage.removeItem(this.userKey);
    this.userSubject.next(null);
  }

  /** Obtiene el access token actual */
  getAccessToken(): string | null {
    return localStorage.getItem(this.tokenKey);
  }

  /** Obtiene el refresh token */
  getRefreshToken(): string | null {
    return localStorage.getItem(this.refreshKey);
  }

  /** Obtiene el usuario almacenado */
  getStoredUser(): UserInfo | null {
    const stored = localStorage.getItem(this.userKey);
    return stored ? JSON.parse(stored) : null;
  }

  /** Verifica si hay sesión activa */
  isLoggedIn(): boolean {
    return !!this.getAccessToken();
  }

  // ── Privado ──

  private exchangeToken(idToken: string, provider: string): Observable<AuthResponse> {
    const endpoint = provider === 'Google' ? 'google' : 'microsoft';
    return this.http
      .post<AuthResponse>(`${this.apiUrl}/auth/${endpoint}`, {
        idToken,
        provider,
      })
      .pipe(tap((res) => this.storeSession(res)));
  }

  private storeSession(res: AuthResponse): void {
    localStorage.setItem(this.tokenKey, res.accessToken);
    localStorage.setItem(this.refreshKey, res.refreshToken);
    localStorage.setItem(this.userKey, JSON.stringify(res.user));
    this.userSubject.next(res.user);
  }
}
